### DISCUSSION TITLE
Parallel execution - Selenium C# NUnit

### DISCUSSION THREAD

**Discussion by ivanlgarcia28** [nunit/nunit#4634] [Q&A] [answered]
Hello guys, hope youre ok!
I'm working on a project with Selenium, C# and NUnit. I'm trying to execute in parallel my tests. It actually works but the chromedriver instances are overlapping itself with each other. For example: both test has microsoft log in page, the first chromedriver instance opens with this page but leave the fields empty, and the second instance types the email twice.. How can I do to get one instance or session id different on this execution? How can tests can be independent? 
Thank you so much, I hope someone can help me..

Regards!

**Comment by stevenaw** [comment]
Is there a small code sample you can share so we can see what you've tried, along with the version of NUnit Framework you're using?

Without knowing how you've setup your project, and without much internal knowledge of selenium, a few general tips I can give which may help:

Try focusing on parallelizing just a small part of the suite first, perhaps just on a single class with multiple contained tests. If your tests are setup in a way which shares state (such as an instance of the Selenium driver stored at class-level) then your tests may be failing if two tests running at the same time are issuing commands to the same instance. This can be avoided in a few ways, but the simplest is to apply the `FixtureLifeCycle(LifeCycle.InstancePerTestCase)]` attribute on the test class to ensure that each test works on a different instance.

**Reply by ivanlgarcia28** [reply]
Hi Steve! Thank you for your answer! 
Here it is a small structure of my project.. I think it can be a little confused how I'm using NUnit tags on each class.. Hope you can figure it what I'm doing wrong here, thank you again!
In my [Test] tag, i added Parallelizable, but on my TestBase, where I have my setup, I added [TestFixture] and [FixtureLifeCycle(LifeCycle.InstancePerTestCase)]
Also my properties class is not visible but I added [assembly: LevelOfParallelism(4)] and Parallelizable(ParallelScope.All)
https://github.com/ivanlgarcia28/SeleniumNUnit

**Reply by stevenaw** [reply, marked as answer]
Happy to help @ivanlgarcia28 .
Based on your current setup you are storing the webdriver instance in a static variable which means it will be shared with all the tests that access it. If you're running into issues with this setup only when your tests run in parallel then it may help to have an instance of the `IWebDriver` class per test. Perhaps something like below would help ensure each test works on a different instance of Selenium so that operations from one test don't interfere with another running at the same time. 

I'd recommend also updating to the latest versions of `Selenium.WebDriver` if you can to ensure you're using any of the latest bugfixes. Your sample uses 4.14 but the latest is 4.18. If you continue to hit issues then it might also be worth reaching out to the Selenium team at https://github.com/SeleniumHQ/selenium for advice as they may have a better sense of what patterns may work to run Selenium in parallel

```
[TestFixture]
[FixtureLifeCycle(LifeCycle.InstancePerTestCase)]
[Parallelizable]
public abstract class TestBase
{
    private readonly IWebDriver Browser { get; set; }

    [SetUp]
    public void SetUp()
    {
        ChromeOptions chromeOptions = new ChromeOptions();
        chromeOptions.AddArguments(new List<string>()
        {
            "--headless",
            "--incognito"
        });
        Browser = new ChromeDriver(chromeOptions);

        Browser.Manage().Window.Maximize();
    }
    [TearDown]
    public void TearDown()
    {
        Browser.Quit();
        Browser.Dispose();
    }
}

**Reply by ivanlgarcia28** [reply]
I think that was the answer! Unfortunately, I'm not able to do the implementation on my project because I will have to modify all my project structure not only the static driver. Appreciate a lot your help! Thank you so much!

**Comment by lakshmi-developer** [comment]
Hi, Could you please help anyone, I want to setup a parllelscope dynamically. declare parlleliza is true and parllelscope in app settings. i want to setup parllelscope dynamically. but i couldn't. i am using below code
 var scope = ConfigProvider.GetConfiguration()["ParallelScope"];
typeof(NUnit.Framework.ParallelizableAttribute)
    .GetConstructor(new Type[] { typeof(ParallelScope) })
    .Invoke(new object[] { Scope });
